### PR TITLE
Spotlight store database (store.db) added LZ4 support #3609

### PR DESCRIPTION
**Pull request by joachimmetz**
Spotlight store database (store.db) added LZ4 support #3609

### CODE REVIEW

File: plaso/parsers/spotlight_storedb.py [hunk 1]
```diff
@@ -896,19 +899,50 @@ def _ReadRecordPage(self, file_object, file_offset):
...
+    if page_header.property_table_type not in (0x00000009, 0x00001009):
       raise errors.ParseError(
           'Unsupported property table type: 0x{0:08x}'.format(
               page_header.property_table_type))
 
     page_data = file_object.read(page_header.page_size - bytes_read)
 
+    file_offset += bytes_read
+
     if page_header.uncompressed_page_size > 0:
```
Comment (Onager) [inline plaso/parsers/spotlight_storedb.py:911]: Pleased unnest/simplify. eg:

```
if page_header.uncompressed_page_size <= 0:
  raise errors.ParseError('Unsupported compression type')

compressed_page_data = page_data
...
``

File: plaso/parsers/spotlight_storedb.py [hunk 2]
```diff
@@ -896,19 +899,50 @@ def _ReadRecordPage(self, file_object, file_offset):
...
+        page_data = lz4.block.decompress(
+            compressed_page_data[12:end_of_compressed_data_offset],
+            uncompressed_size=lz4_block_header.uncompressed_data_size)
+
+        end_of_compressed_data_identifier = compressed_page_data[
+            end_of_compressed_data_offset:end_of_compressed_data_offset + 4]
+
+        if end_of_compressed_data_identifier != b'bv4$':
+          raise errors.ParseError(
+              'Unsupported LZ4 end of compressed data marker')
```
Comment (Onager) [inline plaso/parsers/spotlight_storedb.py:942]: Please include the value of end_of_compressed_data_identifier in the error message

**Review by Onager — CHANGES_REQUESTED** [review]

File: plaso/parsers/spotlight_storedb.py [hunk 3]
```diff
@@ -896,19 +899,50 @@ def _ReadRecordPage(self, file_object, file_offset):
...
-      # TODO: add support for other compression types.
-      if page_data[0] != 0x78:
-        raise errors.ParseError('Unsupported compression type')
+      compressed_page_data = page_data
+
+      if (page_header.property_table_type == 0x00000009 and
+          compressed_page_data[0] == 0x78):
+        page_data = zlib.decompress(compressed_page_data)
+
+      elif (page_header.property_table_type == 0x00001009 and
```
[the remark this answers is not in the document]
Comment (joachimmetz) [reply plaso/parsers/spotlight_storedb.py:918]: done

File: plaso/parsers/spotlight_storedb.py [hunk 4]
```diff
@@ -896,19 +899,50 @@ def _ReadRecordPage(self, file_object, file_offset):
...
+            compressed_page_data[0:4] == b'bv41'):
+        data_type_map = self._GetDataTypeMap(
```
[the remark this answers is not in the document]
Comment (joachimmetz) [reply plaso/parsers/spotlight_storedb.py:920]: done

File: plaso/parsers/spotlight_storedb.py [hunk 5]
```diff
@@ -896,19 +899,50 @@ def _ReadRecordPage(self, file_object, file_offset):
...
-      page_data = zlib.decompress(page_data)
+        try:
+          lz4_block_header = data_type_map.MapByteStream(compressed_page_data)
+        except dtfabric_errors.MappingError as exception:
+          raise errors.ParseError((
+              'Unable to map LZ4 bock header at offset: 0x{0:08x} with error: '
+              '{1!s}').format(file_offset, exception))
+
+        end_of_compressed_data_offset = (
+            12 + lz4_block_header.compressed_data_size)
```
[the remark this answers is not in the document]
Comment (joachimmetz) [reply plaso/parsers/spotlight_storedb.py:931]: done

File: plaso/parsers/spotlight_storedb.py (hunk 2 above)
Comment (joachimmetz) [reply plaso/parsers/spotlight_storedb.py:942]: 4 bytes of random data is not going to tell much. I'll add the offset instead.

File: plaso/parsers/spotlight_storedb.py (hunk 1 above)
Comment (joachimmetz) [reply plaso/parsers/spotlight_storedb.py:911]: so the data can be uncompressed as well, moving the lz4 decompression to a separate function helps improve readability.

File: plaso/parsers/spotlight_storedb.py [hunk 6]
```diff
@@ -879,6 +882,47 @@ def _ReadRecordHeader(self, data, page_data_offset):
...
+      ParseError: if the page data cannot be decompressed.
+    """
+    data_type_map = self._GetDataTypeMap(
+        'spotlight_store_db_lz4_block_header')
+
+    try:
+      lz4_block_header = data_type_map.MapByteStream(compressed_page_data)
+    except dtfabric_errors.MappingError as exception:
+      raise errors.ParseError((
+          'Unable to map LZ4 bock header at offset: 0x{0:08x} with error: '
```
Comment (Onager) [inline plaso/parsers/spotlight_storedb.py:905]: Typo - should be "block" not "bock"

**Review by Onager — APPROVED** [review]
Please fix typo before merging

File: plaso/parsers/spotlight_storedb.py (hunk 6 above)
Comment (joachimmetz) [reply plaso/parsers/spotlight_storedb.py]: done

**Pull request merged**